lib/charms/opensearch/v0/opensearch_plugins.py

```diff
@@ -0,0 +1,219 @@
+
+
+For a new plugin, add the plugin to the list of "ConfigExposedPlugins" available
+in opensearch_plugin_manager.py and override the abstract OpenSearchPlugin.
+
+Optionally, PluginState can also be overridden to add more status.
+
+Add a new configuration in the config.yaml with "plugin_" as prefix to its name.
+Add the corresponding config-name to the ConfigExposedPlugins.
+
+If a given plugin depends on a relation, e.g. repository-s3, then add relation name
+as well. For example:
+    ConfigExposedPlugins = {
+        ...
+        "opensearch-knn": {
+            "class": OpenSearchPlugin,
+            "config-name": "plugin_opensearch_knn",
+            "relation-name": ""
+        },
+    }
+"""
+
+import logging
+import os
+from abc import abstractmethod
+from typing import Any, Dict, List, Optional
+
+from charms.opensearch.v0.helper_enums import BaseStrEnum
+from charms.opensearch.v0.opensearch_exceptions import OpenSearchPluginError
+from jproperties import Properties
+
+# The unique Charmhub library identifier, never change it
+LIBID = "3b05456c6e304680b4af8e20dae246a2"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+logger = logging.getLogger(__name__)
+
+
+class PluginState(BaseStrEnum):
+    """Enum for the states possible in plugins' lifecycle."""
+
+    MISSING = "missing"
+    AVAILABLE = "available"
+    ENABLED = "enabled"
+    WAITING_FOR_UPGRADE = "waiting-for-upgrade"
+
+
+class OpenSearchPlugin:
+    """Abstract class describing an OpenSearch plugin."""
+
+    PLUGIN_PROPERTIES = "plugin-descriptor.properties"
+    CONFIG_YML = "opensearch.yml"
+
+    def __init__(
+        self,
+        plugins_path: str,
+        os_version: Optional[str] = "",
+        relation_data: Optional[Dict[str, Any]] = {},
+    ):
+        """Creates the OpenSearchPlugin object."""
+        self._plugins_path = plugins_path
+        self._os_version = os_version
+        self._relation_data = relation_data
+        self._properties = Properties()
+        self._depends_on = []
+
+    @property
+    def version(self) -> str:
+        """Returns the current version of the plugin."""
+        plugin_props_path = os.path.join(f"{self._plugins_path}", f"{self.PLUGIN_PROPERTIES}")
+        try:
+            with open(plugin_props_path) as f:
+                self._properties.load(f.read())
+        except FileNotFoundError:
+            raise OpenSearchPluginError(f"Plugin is missing - not found: {plugin_props_path}")
+        except PermissionError:
+            raise OpenSearchPluginError(
+                f"Plugin incorrectly installed - permission denied: {plugin_props_path}"
+            )
+        except Exception:
+            raise OpenSearchPluginError("Plugin Error - Unknown reason")
+        return self._properties._properties["version"]
```

Comment:
Personally, this is related to the ```def version(self)``` method, which is very internal to the plugin infrastructure. I would rather have it returning exclusively OpenSearchPluginError instead of different excepts.